### PR TITLE
consensus: add the DoneAddingBlock tracer event

### LINKED ISSUE #3382: Add new tracer event that reports a block's propagation delay within our node
@karknu and @njd42 pointed out that a a helpful diagnostic metric on running nodes would be (statistics about) the delay between receiving a block and when we began offering it to peers. In some sense, this is our node's internal contribution to the that block's propagation delay (at least for p2p connections whose shortest path involves us).

This Issue is to create a tracer event that includes that specific delay along with the block's metadata and likely additional information about the corresponding chain selection event (eg if it was a roll back, eg sizes of other blocks).

There are some open questions, such as: Would it be ok to only emit this event for some block when that block becomes our _tip_? Recall that a block can be on our chain without having ever been our tip (eg if it's in the prefix of short fork that later wins out). Only emitting the event for our tip would be quite a bit easier to implement, eg (boils down to adding an ingress timestamp to the add-block queue, I think).

### PR DESCRIPTION
**Pull request by nfrisby**
Closes #3382.

Single commit PR, please see its message.

### CODE REVIEW

**Comment by nfrisby** [comment]
Chatted with Karl on Slack.

- We need an absolute time for adoption, so that we can compare it to not only when the block entered the queue but also when the block was forged (ie when it labelled slot began).

- We do not need to worry about out-of-order block reception. EG It's infrequent enough that it's fine to ignore such events in our metrics.

**Comment by nfrisby** [comment]
I rebased and also added a `(Time, UTCTime)` argument as promised in my previous message.

The `UTCTime` part in particular required "polluting" parts of our codebase. Specifically, I had to undo parts of the commit Edsko wrote (d0bd293107c2fc901d70cb6ffb087f5d7632c65c) when he was consolidating all the uses of time out of the Consensus Layer in general and isolating it only in the HFC logic (which was not a simple task and we don't want to let time creep back in).

One compromise that avoids this pollution: make this third argument just a `Time`. Then, if the event processing framework would provide the time of the event (both as `Time` and as `UTCTime`) to the functions that process events (eg render them as log messages), then the event processors could determine the `UTCTime` portion of the third time argument in this `DoneAddingBlock` event. Note that the event processing framework already has this time, since that's what it writes as the prefix of the log line! (This idea is part way towards the "ideal" discussed in my previous comment: https://github.com/input-output-hk/ouroboros-network/pull/3427#issuecomment-946855525)

**Comment by nfrisby** [comment]
(The following is adapted from a Slack message I sent to Karl, Serge, and Jordan. It largely summarizes the state of this PR. If this PR were merged today, then the downstream consequences are reflected in the `cardano-node` commit `6b3bc25f6`, discussed below. If those stakeholders are content with that, then this PR is mergable without further external negotations, other than maybe name bikeshedding, etc. cc @dnadales )

I've pushed up a few commits.

These two are in `ouroboros-network`. They're the same, except `31379a1d6` was rebased (no conflicts) onto the same `ouroboros-network` commit that Karl had been using on his `cardano-node` branch.

```981d48017 - (HEAD -> nfrisby/issue-3382-trace-block-processing-time) consensus: add the DoneAddingBlock tracer event (2 seconds ago) <Nicolas Frisby>
afc8db599 - (origin/master, origin/HEAD) Merge #3455 (28 hours ago) <iohk-bors[bot]>

31379a1d6 - (origin/tmp/nfrisby/issue-3382-demo) consensus: add the DoneAddingBlock tracer event (20 minutes ago) <Nicolas Frisby>
aa7bc0877 - (origin/jc/benchmark/1-29-0-with-time) Max known PV is now the Alonzo PV. (9 weeks ago) <Nicholas Clarke>```

This one extends Karl's `cardano-node` branch in order to use the changes I just pushed to `ouroboros-network` (ie `31379a1d6`).

```6b3bc25f6 - (HEAD -> nfrisby/block_adoption_delay, origin/nfrisby/block_adoption_delay) WIP: update for latest ouroboros-network PR (16 minutes ago) <Nicolas Frisby>
8be81bbc1 - (origin/karknu/block_adoption_delay) WIP: Block adoption metrics (5 weeks ago) <Karl Knutsson>```

Thanks to Jordan for suggesting the custom `trTransformer` instance.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs
Comment (nfrisby) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs:309]: This split out remains a TODO.

Also, I stubbornly labored to avoid a `MonadTime` constraint on this function. The `SupportsNode` constraint is similarly warty: this function doesn't need that constraint except to emit this event. That seems like poor separation of concerns (as evidence eg by this PR needing the `TestBlock` instance...).

If we paid the price to complicate the `Tracer` event types, then we could have a top-level type-changing tracer transformer that allowed for the `Tracer` here to know how to do this translation instead of this function itself knowing how. For downstream users, they'd just see an extra type parameter on the tracer events that Consensus emits, but it'd have a fix value by the time they care about it.

If we were to do this, then the `MonadTime` issue could be handled in the same way. Maybe it's preferable in the long-term to the current solution on this PR.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs:888]: I haven't given thought to where this type declaration should actually be defined. It's an annoying shortcoming of GHC Haskell that we need these. (We could get around it, but only if we established a fixed interface boundary on `Tracer` events, and used eg pattern synonyms to hide it. Which we haven't bothered with so far.)

**Comment by nfrisby** [comment]
My previous top-level comment explains that I hope this PR is now only waiting on internal review. I just did some in situ comments about the topics I expect internal review to resolve.

**Review by dnadales — APPROVED** [review]

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs
Comment (dnadales) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs:309]: I went through the PR once more, and I approved it as it looks good to me. My only concern is the additional 13 lines of code in the `addBlockSync` function that are there only for the sake of logging the `DoneAddingBlock` event. I wonder if this pollutes the logic of this function, or whether, on the contrary, describing how the forging time of a block can be calculated is pertinent to `addBlockSync`. Maybe we can talk about this in a call.

**Comment by nfrisby** [comment]
The status here is that I'm not happy with the complication in `ChainSel`. Damian noted this as well in one of his comments.

On one of Javier's recent PRs, we came up with a new lightweight scheme for handling this sort of `Tracer` event where the code emitting the tracer doesn't/shouldn't have all the information the `Tracer` carries while the more outer layers of code do (function spaces). Maybe we could use that here.

**Comment by nfrisby** [comment]
I'm hoping PR #3540 supersedes this one. It's using a few lightweight types to maintain the desired separation of concerns in a more granular way, which allows for the `DoneAddingBlock` event to be simple. In particular, there's no need for the post-processing pass to get times out

So: if PR 3540 merges, we can Close this one without merging it. Until then, something similar to this would be my fallback PR.

**Comment by nfrisby** [comment]
Closing as stale.

**Pull request closed without merging**
